### DISCUSSION TITLE
Defunct processes on macOS when exiting a shell with `exit` or `Ctrl-d`

### DISCUSSION THREAD

**Discussion by creechy** [ghostty-org/ghostty#3596] [Q&A] [answered]
I don't think this is quite a regression of Issue https://github.com/ghostty-org/ghostty/issues/2273, but I see defunct/zombie processes on macOS when exiting the shell with `exit` or `Ctrl-d`. It _does not_ happen when I click the `X` on the shell tab/title bar. I believe, like the other issue, this is the `login` process, though.

My environment is macOS Sequoia 15.2, I have built my own `bash` shell with

```
./configure --prefix=/opt/local
```

Capturing some logging, I don't see much, when I exit a shell, not sure what I should expect though

```
2024-12-27 10:08:21.734914-0800 0x41f8ae   Info        0x0                  83542  0    ghostty: [com.mitchellh.ghostty:io_exec] read thread got quit signal
2024-12-27 10:08:21.738014-0800 0x41f477   Info        0x0                  83542  0    ghostty: [com.mitchellh.ghostty:surface] surface closed addr=138028000
```

**Comment by basile-henry** [comment]
I have the same problem on Ubuntu 24.04: `Ctrl-D` leaves the ghostty process running in the background (I can see it in htop/system monitor) whereas `X` properly closes it.

**Reply by creechy** [reply]
That might be a little different. When I open a few tabs with `Command-t` and then exit the shells with `Ctrl-d`, I get a bunch of defunct processes 
```
mock@paget[4]: ps -aef | grep defunct
    0 98168 95941   0  2:42PM ??         0:00.00 <defunct>
    0 98203 95941   0  2:42PM ??         0:00.00 <defunct>
    0 98238 95941   0  2:42PM ??         0:00.00 <defunct>
    0 98273 95941   0  2:42PM ??         0:00.00 <defunct>
```

These were the `login` processes of the shells that were opened.

**Comment by lukasmalkmus** [comment]
Same here on macOS. I was expecting Ghostty to quit once I `Ctrl-d`.

**Reply by goonzoid** [reply]
If you just want Ghostty to quit when you exit your last running shell, there is this: https://ghostty.org/docs/config/reference#quit-after-last-window-closed

**Comment by nklmilojevic** [comment]
Also seeing this on macOS.

**Comment by goonzoid** [comment, marked as answer]
Spent some time debugging this today. See https://github.com/ghostty-org/ghostty/issues/4554 for the gory details and a link to a potential fix on my fork.

**Reply by 00-kat** [reply]
Looks like #4554 was closed, so I'm assuming this is fixed now.

**Reply by creechy** [reply]
Its fixed for me. I have not seens defunct/zombie processes since I upgraded to 1.1.0

**Reply by jparise** [reply]
#9760 is also related.
